### PR TITLE
[SDFAB-1148] Support canonical byte string in PTF tests

### PR DESCRIPTION
**Pull request by Yi-Tseng**
This PR changed the way we generate P4Runtime messages, we always canonicalized the value when we are setting values to the actual P4Runtime message in order to comply with the P4Runtime spec.

### CODE REVIEW

File: ptf/tests/common/base_test.py [hunk 1]
```diff
@@ -117,26 +117,10 @@ def get_controller_packet_metadata(p4info, meta_type, name):
...
-    if bitwidth <= 0:
-        raise ValueError("bitwidth must be a positive integer.")
-    if input is None:
-        raise ValueError("input cannot be of NoneType.")
-
-    byte_width = (
-        bitwidth + 7
-    ) // 8  # use integer division to avoid floating point rounding errors.
-    return input.rjust(byte_width, b"\0")  # right padding <-> BigEndian
+def canonicalized(byte_string):
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:120]: shorter and easier to pronounce
```suggestion
def canonical(byte_string):
```

File: ptf/tests/common/base_test.py [hunk 2]
```diff
@@ -381,17 +365,10 @@ def verify_packet_in(self, exp_pkt, exp_in_port, timeout=2):
...
-                rx_in_port_ = de_canonicalize_bytes(
-                    pkt_in_ig_port_bitwidth, rx_in_port_
-                )
-                rx_inport = struct.unpack("!H", rx_in_port_)[0]
+            rx_inport = 0
+            for b in rx_in_port_:
+                rx_inport <<= 8
+                rx_inport += b
 
             if exp_in_port != rx_inport:
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:373]: Is this equivalent? If yes, would this be easier to read?
```suggestion
            exp_in_port_ = stringify(exp_in_port)
            if canonical(exp_in_port_) != canonical(rx_inport_):
```

**Review by ccascone — COMMENTED** [review]
Can we also remove the bitwidth parameter from `stringify`? That should be useless now, right?

File: ptf/tests/common/base_test.py (hunk 2 above)
Comment (Yi-Tseng) [reply ptf/tests/common/base_test.py]: We can actually get the value with `int.from_bytes` which I think it is simpler

**Comment by Yi-Tseng** [comment]
> Can we also remove the bitwidth parameter from `stringify`? That should be useless now, right?

There is a case we still need is that we need to make sure the ternary match value size is the same as the mask size.
But I believe that we can just add zeros before we check it
https://github.com/stratum/fabric-tna/blob/main/ptf/tests/common/base_test.py#L618

File: ptf/tests/common/base_test.py [hunk 3]
```diff
@@ -42,8 +41,12 @@
 RPC_TIMEOUT = 10  # used when sending Write/Read requests.
 
 # Convert integer (with length) to binary byte string
-def stringify(n, length):
-    return n.to_bytes(length, byteorder="big")
+def stringify(n):
+    byte_length = (n.bit_length() + 7) // 8
+    # We will get 0 `n` is zero, in this case we need to set byte length
+    # to 1 so the final result will be b"\x00".
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:47]:
```suggestion
    # We will get 0  if `n` is zero, in this case we need to set byte length
    # to 1 so the final result will be b"\x00".
```

File: ptf/tests/common/base_test.py [hunk 4]
```diff
@@ -582,18 +575,22 @@ def add_to(self, mf_id, mk, bitwidth):
             mf.lpm.prefix_len = self.pLen
             mf.lpm.value = b""
 
+            # De-canonicalized before removing don't-care bits.
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:578]:
```suggestion
            # De-canonicalize before zeroing don't-care bits.
```

File: ptf/tests/common/base_test.py [hunk 5]
```diff
@@ -602,6 +599,9 @@ def __init__(self, mf_name, v, mask):
             self.mask = mask
 
         def add_to(self, mf_id, mk, bitwidth):
+            # De-canonicalized before removing don't-care bits.
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:602]:
```suggestion
            # De-canonicalize before zeroing don't-care bits.
```

File: ptf/tests/common/base_test.py [hunk 6]
```diff
@@ -117,13 +120,10 @@ def get_controller_packet_metadata(p4info, meta_type, name):
                         return m
 
 
-def de_canonicalize_bytes(bitwidth: int, input: bytes):
+def de_canonical(bitwidth: int, input: bytes):
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:123]: Let's call this for what it is :)
```suggestion
def pad(bitwidth: int, input: bytes):
```

File: ptf/tests/common/base_test.py [hunk 7]
```diff
@@ -117,13 +120,10 @@ def get_controller_packet_metadata(p4info, meta_type, name):
...
-    This method adds a padding to the 'input' param.
-    Needed for bmv2 since it uses Canonical Bytestrings: this representation
-    trims the data to the lowest amount of bytes needed for that particular value
-    (e.g. 0x0 for PacketIn.ingress_port will be interpreted by Stratum bmv2 using 1 byte, instead of 9 bits,
-    as declared in header.p4)
+    This method adds padding zeros to the 'input' param.
+
     :param bitwidth: the desired size of input.
     :param input: the byte string to be padded.
     :return: padded input with bytes such that: len(bin(input)) >= bitwidth.
```
Comment (ccascone) [inline ptf/tests/common/base_test.py:129]: Not related to this PR but doesn't hurt to fix here
```suggestion
    :param bitwidth: the desired size of the output in bits.
    :param input: the byte string to be padded.
    :return: input padded with zero bytes such that: len(bin(input)) = bytewidth.
```

**Review by ccascone — APPROVED** [review]
Left only minor comments. Would be good to fix them, but I don't need to review again. Please merge when ready.

**Pull request merged**